File: android/ml/orchestrator/Orchestrator.kt
package com.vachak.engine.orchestrator

import com.vachak.engine.*

/**
 * Orchestrator — request lifecycle (PHASE 9A). Kotlin wiring that binds the
 * shared/orchestrator state machine to the existing EngineContracts.
 *
 * Sequential only (RAM budget): ASR -> MT -> TTS -> play. Never parallelize.
 * Uses EngineResult<T> uniformly; EngineError.MODEL_NOT_LOADED is the signal for
 * PROMPT_LANGUAGE_PACK (see ErrorPolicy).
 *
 * Components come from the EngineProvider (mock in dev, sherpa-onnx /
 * IndicTrans2 in production). This file only sequences them and applies the
 * graceful-degradation policy; it contains no model code.
 */
enum class PipelineStage { CAPTURE, VAD, ASR, CONTEXT, TRANSLATE, VALIDATE, SYNTHESIZE, SPEAK, DONE }

enum class RequestStatus { OK, DEGRADED, FAILED }

data class StageResult(val stage: String, val ok: Boolean, val ms: Long, val error: String? = null, val action: String? = null)

data class PipelineResult(
    val status: RequestStatus,
    val transcript: String? = null,
    val translation: String? = null,
    val presentText: String? = null,
    val presentAudio: ShortArray? = null,
    val promptLanguagePack: Boolean = false,
    val error: String? = null,
    val stages: List<StageResult> = emptyList(),
    val totalMs: Long = 0L,
)

/**
 * Local contract for terminology validation (not in EngineContracts). Swap with
 * a glossary-backed implementation; on failure the orchestrator shows the source.
 */
interface TerminologyValidator {
    fun validate(text: String, context: ClassroomState): ValidationOutcome
    data class ValidationOutcome(val ok: Boolean, val normalized: String, val issues: List<String>)
}

/**
 * Local contract for audio playback (EngineContracts.TTSEngine returns PCM;
 * AudioTrack owns playback on device).
 */
interface Speaker {
    fun play(pcm16: ShortArray)
}

class Orchestrator(
    private val asr: ASREngine,
    private val nmt: TranslationEngine,
    private val tts: TTSEngine,
    private val validator: TerminologyValidator? = null,
    private val speaker: Speaker? = null,
    private val packManager: LanguagePackManager? = null,
    private val state: ClassroomState = ClassroomState(),
    private val policy: ErrorPolicy = ErrorPolicy(),
    private val budgetMs: Long = LatencyBudget.TOTAL_MS,
) {
    fun processUtterance(pcm16: ShortArray, sampleRateHz: Int): PipelineResult {
        val t0 = System.nanoTime()
        val stages = mutableListOf<StageResult>()

        // 1. ASR with retry
        var transcript: String? = null
        var asrOk = false
        var lastErr: String? = null
        for (attempt in 0..policy.asrRetries) {
            when (val r = asr.transcribe(pcm16, sampleRateHz)) {
                is EngineResult.Ok -> { transcript = r.value; asrOk = true
                    stages += StageResult(PipelineStage.ASR.name, true, 0); break }
                is EngineResult.Err -> {
                    lastErr = r.message
                    stages += StageResult(PipelineStage.ASR.name, false, 0, r.message)
                    if (r.code == EngineError.MODEL_NOT_LOADED)
                        return finish(RequestStatus.FAILED, stages, t0, promptPack = true, error = policy.modelMessage)
                    val (deg, msg) = policy.decide("asr", attempt, false)
                    if (deg != Degradation.RETRY) return finish(RequestStatus.FAILED, stages, t0, error = msg)
                }
            }
        }
        if (!asrOk) return finish(RequestStatus.FAILED, stages, t0, error = lastErr)
        state.transcript = transcript

        // 2. Translate (fallback: show source)
        val pair = LanguagePair(state.sourceLanguage, state.targetLanguage)
        val translation = when (val r = nmt.translate(transcript!!, pair)) {
            is EngineResult.Ok -> r.value
            is EngineResult.Err -> {
                if (r.code == EngineError.MODEL_NOT_LOADED)
                    return finish(RequestStatus.FAILED, stages, t0, promptPack = true, error = policy.modelMessage)
                return finish(RequestStatus.DEGRADED, stages, t0, presentText = transcript,
                    error = policy.translateMessage)
            }
        }
        state.translation = translation
        stages += StageResult(PipelineStage.TRANSLATE.name, true, 0)

        // 3. Validate
        if (validator != null) {
            val v = validator.validate(translation, state)
            if (!v.ok)
                return finish(RequestStatus.DEGRADED, stages, t0, presentText = transcript,
                    error = "terminology validation failed")
            state.translation = v.normalized
        }

        // 4. Synthesize (fallback: show translation)
        val audio = when (val r = tts.synthesize(translation, state.targetLanguage)) {
            is EngineResult.Ok -> r.value
            is EngineResult.Err -> {
                if (r.code == EngineError.MODEL_NOT_LOADED)
                    return finish(RequestStatus.FAILED, stages, t0, promptPack = true, error = policy.modelMessage)
                return finish(RequestStatus.DEGRADED, stages, t0, presentText = translation,
                    error = policy.ttsMessage)
            }
        }
        stages += StageResult(PipelineStage.SYNTHESIZE.name, true, 0)

        // 5. Speak
        speaker?.play(audio)
        return finish(RequestStatus.OK, stages, t0, presentAudio = audio, transcript = transcript, translation = translation)
    }

    private fun finish(
        status: RequestStatus, stages: List<StageResult>, t0: Long,
        presentText: String? = null, presentAudio: ShortArray? = null,
        transcript: String? = null, translation: String? = null,
        error: String? = null, promptPack: Boolean = false,
    ): PipelineResult {
        state.lastStage = PipelineStage.DONE.name
        state.lastStatus = status.name
        state.lastError = error
        state.lastAction = if (promptPack) "prompt_language_pack" else null
        return PipelineResult(
            status = status, transcript = transcript ?: state.transcript,
            translation = translation ?: state.translation,
            presentText = presentText, presentAudio = presentAudio,
            promptLanguagePack = promptPack, error = error, stages = stages,
            totalMs = (System.nanoTime() - t0) / 1_000_000,
        )
    }
}

File: shared/orchestrator/orchestrator.py
"""Orchestrator — owns the full request lifecycle (PHASE 9A).

This is the language-agnostic reference implementation. It drives the sequential
voice pipeline, measures per-stage latency, and applies the graceful error
policy (PHASE 9C). The Kotlin wiring in android/ml/orchestrator mirrors it
against the existing EngineContracts types.

Design notes:
  * Sequential execution only (RAM budget on 2GB devices). Never parallelize
    ASR/MT/TTS.
  * Every stage is timed with perf_counter so the benchmark harness can reuse
    the same code path on-device.
  * A `ModelUnavailableError` from any component triggers the language-pack
    action rather than a generic failure.
"""
from __future__ import annotations

import time
from dataclasses import dataclass, field
from enum import Enum
from typing import Any, Optional

from .classroom_state import ClassroomState
from .errors import Degradation, ErrorPolicy, decide
from .interfaces import (
    ASR,
    ContextEngine,
    ModelUnavailableError,
    NMT,
    Speaker,
    TTS,
    TerminologyValidator,
    VAD,
    VadResult,
)


class PipelineStage(str, Enum):
    CAPTURE = "capture"
    VAD = "vad"
    ASR = "asr"
    CONTEXT = "context"
    TRANSLATE = "translate"
    VALIDATE = "validate"
    SYNTHESIZE = "synthesize"
    SPEAK = "speak"
    DONE = "done"


class RequestStatus(str, Enum):
    OK = "ok"
    DEGRADED = "degraded"        # succeeded via fallback (e.g. showed source text)
    FAILED = "failed"


@dataclass
class StageResult:
    stage: str
    ok: bool
    ms: float
    error: Optional[str] = None
    action: Optional[str] = None


@dataclass
class PipelineResult:
    status: RequestStatus
    transcript: Optional[str] = None
    translation: Optional[str] = None
    audio: Any | None = None
    # What the UI should actually present to the teacher/student.
    present_text: Optional[str] = None
    present_audio: Any | None = None
    prompt_language_pack: bool = False
    error: Optional[str] = None
    stages: list[StageResult] = field(default_factory=list)
    total_ms: float = 0.0

    @property
    def within_budget_ms(self) -> float:
        return self.total_ms


@dataclass
class Components:
    vad: Optional[VAD] = None
    asr: Optional[ASR] = None
    context: Optional[ContextEngine] = None
    nmt: Optional[NMT] = None
    validator: Optional[TerminologyValidator] = None
    tts: Optional[TTS] = None
    speaker: Optional[Speaker] = None


class Orchestrator:
    def __init__(
        self,
        components: Components,
        state: Optional[ClassroomState] = None,
        policy: Optional[ErrorPolicy] = None,
        latency_budget_ms: float = 3000.0,
    ) -> None:
        self.c = components
        self.state = state or ClassroomState()
        self.policy = policy or ErrorPolicy()
        self.latency_budget_ms = latency_budget_ms

    # -- helpers -----------------------------------------------------------
    def _time(self, fn, *args, **kw):
        t0 = time.perf_counter()
        out = fn(*args, **kw)
        return out, (time.perf_counter() - t0) * 1000.0

    def _is_model_unavailable(self, exc: Exception) -> bool:
        return isinstance(exc, ModelUnavailableError)

    # -- main entry --------------------------------------------------------
    def process_utterance(self, audio: Any) -> PipelineResult:
        """Full lifecycle: audio in, audio/text out (or graceful degradation)."""
        stages: list[StageResult] = []
        t0_all = time.perf_counter()

        # 1. VAD
        try:
            seg, ms = self._time(self._run_vad, audio)
        except Exception as exc:  # pragma: no cover - vad rarely fails
            stages.append(StageResult(PipelineStage.VAD, False, 0, str(exc)))
            return self._finalize(RequestStatus.FAILED, stages, t0_all, error=str(exc))
        stages.append(StageResult(PipelineStage.VAD, True, ms))

        # 2. ASR (with retry)
        transcript: Optional[str] = None
        asr_ok = False
        last_err: Optional[str] = None
        for attempt in range(self.policy.asr.retries + 1):
            try:
                transcript, ms = self._time(self.c.asr.transcribe, seg.audio)
                stages.append(StageResult(PipelineStage.ASR, True, ms))
                asr_ok = True
                break
            except Exception as exc:
                last_err = str(exc)
                deg, msg = decide(PipelineStage.ASR, attempt, self.policy,
                                  self._is_model_unavailable(exc))
                stages.append(StageResult(PipelineStage.ASR, False, 0, str(exc), deg.value))
                if deg is Degradation.RETRY:
                    continue
                # model unavailable or out of retries -> fail
                if deg is Degradation.PROMPT_LANGUAGE_PACK:
                    return self._finalize(RequestStatus.FAILED, stages, t0_all,
                                          error=msg, prompt_pack=True)
                return self._finalize(RequestStatus.FAILED, stages, t0_all, error=msg)
        if not asr_ok:
            return self._finalize(RequestStatus.FAILED, stages, t0_all, error=last_err)

        self.state.record(PipelineStage.ASR, status="ok", transcript=transcript)

        # 3. Context
        ctx = self._run_context(transcript)
        stages.append(StageResult(PipelineStage.CONTEXT, True, 0.0))

        # 4. Translate (fallback: show source)
        translation: Optional[str] = None
        try:
            translation, ms = self._time(self._run_translate, transcript, ctx)
            stages.append(StageResult(PipelineStage.TRANSLATE, True, ms))
            self.state.record(PipelineStage.TRANSLATE, status="ok", translation=translation)
        except Exception as exc:
            deg, msg = decide(PipelineStage.TRANSLATE, 0, self.policy,
                              self._is_model_unavailable(exc))
            stages.append(StageResult(PipelineStage.TRANSLATE, False, 0, str(exc), deg.value))
            if deg is Degradation.PROMPT_LANGUAGE_PACK:
                return self._finalize(RequestStatus.FAILED, stages, t0_all,
                                      error=msg, prompt_pack=True)
            # SHOW_SOURCE fallback
            self.state.record(PipelineStage.TRANSLATE, status="degraded", error=msg,
                              action=deg.value)
            return self._finalize(RequestStatus.DEGRADED, stages, t0_all,
                                  transcript=transcript, present_text=transcript, error=msg)

        # 5. Validate
        validation_ok = True
        if self.c.validator is not None:
            try:
                vr, ms = self._time(self.c.validator.validate, translation, ctx)
                stages.append(StageResult(PipelineStage.VALIDATE, vr.ok, ms,
                                          None if vr.ok else "; ".join(vr.issues or [])))
                validation_ok = vr.ok
                translation = vr.normalized
            except Exception as exc:  # validation failure -> show source
                stages.append(StageResult(PipelineStage.VALIDATE, False, 0, str(exc)))
                validation_ok = False
        if not validation_ok:
            self.state.record(PipelineStage.VALIDATE, status="degraded",
                              translation=translation, action="show_source")
            return self._finalize(RequestStatus.DEGRADED, stages, t0_all,
                                  transcript=transcript, present_text=transcript,
                                  error="terminology validation failed")

        # 6. Synthesize (fallback: show translation)
        audio_out: Any | None = None
        try:
            audio_out, ms = self._time(self._run_tts, translation)
            stages.append(StageResult(PipelineStage.SYNTHESIZE, True, ms))
            self.state.record(PipelineStage.SYNTHESIZE, status="ok", audio=audio_out)
        except Exception as exc:
            deg, msg = decide(PipelineStage.SYNTHESIZE, 0, self.policy,
                              self._is_model_unavailable(exc))
            stages.append(StageResult(PipelineStage.SYNTHESIZE, False, 0, str(exc), deg.value))
            if deg is Degradation.PROMPT_LANGUAGE_PACK:
                return self._finalize(RequestStatus.FAILED, stages, t0_all,
                                      error=msg, prompt_pack=True)
            self.state.record(PipelineStage.SYNTHESIZE, status="degraded",
                              translation=translation, action=deg.value)
            return self._finalize(RequestStatus.DEGRADED, stages, t0_all,
                                  transcript=transcript, translation=translation,
                                  present_text=translation, error=msg)

        # 7. Speak
        try:
            _, ms = self._time(self._run_speak, audio_out)
            stages.append(StageResult(PipelineStage.SPEAK, True, ms))
        except Exception as exc:  # speaking failing still leaves audio available
            stages.append(StageResult(PipelineStage.SPEAK, False, 0, str(exc)))
            self.state.record(PipelineStage.SPEAK, status="degraded",
                              translation=translation, audio=audio_out)
            return self._finalize(RequestStatus.DEGRADED, stages, t0_all,
                                  transcript=transcript, translation=translation,
                                  present_audio=audio_out,
                                  error="playback failed; audio ready")

        self.state.record(PipelineStage.DONE, status="ok", audio=audio_out)
        return self._finalize(RequestStatus.OK, stages, t0_all,
                              transcript=transcript, translation=translation,
                              audio=audio_out, present_audio=audio_out)

    # -- stage runners (kept tiny so failures are isolated) -----------------
    def _run_vad(self, audio):
        if self.c.vad is None:
            return VadResult(audio=audio)
        return self.c.vad.detect(audio)

    def _run_context(self, transcript):
        if self.c.context is None:
            return None
        return self.c.context.resolve(transcript)

    def _run_translate(self, transcript, ctx):
        return self.c.nmt.translate(
            transcript, self.state.source_language, self.state.target_language
        )

    def _run_tts(self, translation):
        return self.c.tts.synthesize(translation, self.state.target_language)

    def _run_speak(self, audio_out):
        if self.c.speaker is not None:
            self.c.speaker.play(audio_out)

    # -- finalize ----------------------------------------------------------
    def _finalize(self, status, stages, t0_all, *, transcript=None, translation=None,
                  audio=None, present_text=None, present_audio=None, error=None,
                  prompt_pack=False) -> PipelineResult:
        total = (time.perf_counter() - t0_all) * 1000.0
        res = PipelineResult(
            status=status,
            transcript=transcript if transcript is not None else self.state.transcript,
            translation=translation if translation is not None else self.state.translation,
            audio=audio,
            present_text=present_text,
            present_audio=present_audio,
            prompt_language_pack=prompt_pack,
            error=error,
            stages=stages,
            total_ms=total,
        )
        self.state.record(
            PipelineStage.DONE if status is RequestStatus.OK else "error",
            status=status.value, error=error,
            action="prompt_language_pack" if prompt_pack else None,
        )
        return res


__all__ = [
    "Orchestrator",
    "Components",
    "PipelineStage",
    "RequestStatus",
    "StageResult",
    "PipelineResult",
]
